[BUCash] - Update BUCash to allow BU/BUCash side-by-side w/o interfering with each other's Qt Settings

Justaphf opened the pull request:
Looking for feedback on this implementation.

To allow side-by-side installs of BitcoinUnlimited and BUCash on the same machine, update the Qt App Name used by BUCash.  This will result in different registry location on Windows, or different settings file on Linux/Mac.

1. If BUCash, check to see if settings have already been migrated to the new app name.
2. If migration is needed, perform the full migration from old app name to new app name.
2.a. First check if there was already a BU install and clone these settings.
2.b. If no BU install, but another Satoshi based client installed, clone those settings instead.
3. If a migration occurred (BUCash client only), force the user to re-pick their data directory.  This is needed because side-by-side clients will need to maintain different data directories for BTC main chain vs BCC main chain.

Questions for feedback:
1. Should the migrated settings outright ignore the `strDataDir` setting from the source settings?  Prompting the user to reselect their data directory only works if the settings were just migrated in this session.  If they kill the app after the settings have been migrated, but having not yet selected a directory, they won't be prompted to reselect the next time they start bitcon-qt and the client will just use the migrated data directory (if not overridden via command-line/config file).
2. Are registry settings/configuration for Qt set as part of the Windows/Mac installers? I didn't see anything searching for specific string so I've only updated the bitcoin-qt client code.

TODO:
- Test on Linux/Mac (already tested on Windows)
- Possibly update text on the data directory picker dialog to add a warning for the BUCash client that if the user intends to run side-by-side clients for both the BTC and BCC chains, they need to have different data directories, and this is why the user is being prompted to pick the directory again.

ptschip commented:
@Justaphf great idea...

ptschip commented:
@Justaphf 

I wonder if you should make 2 PR's...one just for windows installation so that users can install both clients and launch them in windows (ie the diff registry keys)  and leave the migration to a separate PR as it's probably more complex and need a lot of testing?

Justaphf commented:
@ptschip Thanks for taking a look.  I will restructure my commits to make it easier to separate out the various sub-functionality.  The reason I had included the migration code as part of this PR is because the code to perform the migration already exists for migrating to the BU registry, it just needed to be tweaked to point towards the BUCash registry instead (and otherwise just having the new registry settings would lose all previous settings where the datadir is really the only setting that needs to be updated).  I added on top of that the pop-up dialog to have the user reconfirm the data directory since legacy and cash chains need separate data directories.

Justaphf committed:
Add separate Qt settings for BUcash

This is so a non-cash client can be run on the same machine as a cach
client without interfering with each other's registry settings.

Justaphf committed:
Use const for legacy org name instead of local string literal

Justaphf committed:
Update QSettings migrate fxn to take app name as input

rather than rely on hard coding the app name inside the function.
This is a precursor to allowing migration of settings from BU to BUcash
where the org name is the same, but the app name is different.

Justaphf committed:
Migrate QT settings for BUcash nodes

In the case where settings were just migrated as part of startup, if
this is a BUcash client, force a re-confirmation of the data directory,
which will be stored in the registry (Windows) or a settings file
(Linux/Mac).  This is because if the user wants to run dual nodes for
BTC and BCC chains on the same machine, it will require separate data
directories.  This only affects bitcoin-qt.

Justaphf commented:
Restructured commits again to make them more easily separable, if necessary.
* Commits 1 & 2 are clean-up and improvement of the existing `QSettings` migration functionality.
  * These are not BUCash specific and would be good to have on `dev` branch too.
* Commit 3 sets up the separate `QSettings` for BUCash without attempting migration from old settings
* Commit 4 adds in the migration functionality. BU->BUCash otherwise Legacy->BUCash

Thinking it may be better to rebase on top of `dev` rather than `BitcoinCash` branches as all cash specific functionality is protected by preprocessor statements.  Alternately I could create a new PR just for commits 1 & 2 on `dev` and then this PR would need to rebase after `dev` merges to `BitcoinCash` branch for commits 3 & 4 if we don't want to pollute `dev` with `BITCOIN_CASH` preprocessor statements.

ptschip commented:
@Justaphf just starting to take a look but I think you are right in that this should be rebased on dev rather than just the cash branch.

ptschip commented on src/qt/bitcoin.cpp:
```diff
@@ -702,9 +702,20 @@ int main(int argc, char *argv[])
-    // TODO: Add migration code which first tries to move from BU to BUcash settings,
-    //       or if there weren't pre-existing BU settings, try to migrate from a legacy settings
+    bool bMigrated = false;
+    // For BUCash, first try to migrate from BTC BU settings
```
nit, although i don't think its formally specified anywhere, in bitcoin we use an f prefix instead of b to indicate a bool.

ptschip commented:
@Justaphf nit:  when i first launch the welcome screen says at least 80GB of data will be stored as blockchain data...I think this message is a bit dated as far as size and I wonder whether we should just take out the reference to the 80GB since whatever number we put in going to need updating on occasion.  I think most node operators are well aware of the size of data needed...also this number doesn't take into account a pruned node and 80GB could be much higher than the user would allow?

ptschip commented:
@Justaphf  test results show that this process generally works but the one thing that stuck out for me is that the bitcoin.conf settings are not migrated over.  So now what happens if the user decides to not keep the lecacy chain anymore and deletes the old datadir later...now the orig bitcoin.conf is gone but the user will likely think his setting have been preserved.  IMO, we should also migrate the bitcoin.conf settings over, at the time of update, as well.

ptschip commented:
@Justaphf i'm not sure now whether this edge case of the bitcoin.conf not getting carried over is really a valid one...It only happened to me because I created an empty BitcoinCash datadir for testing the migration.   But in reality, the user will already have a BitcoinCash datadir with the blockchain already or partially syncd, and with a bitcoin.conf file present.  Futhermore if they do a new install of BitcoinCash they should have everything they need in the right data folder.  So I don't think this edge case really need to be solved, thoughts?

Justaphf commented:
@ptschip Thanks for taking a look at this.  I'll rebase on dev and fix the boolean nit tonight or tomorrow.

For the other issues
* As part of this PR I haven't touched the information displayed on the select data dir dialog (just caused it to display again if migrating from non-cash to cash).  It's probably a good idea to change that message, but seems like a separate PR unless we want to add wording specific to migration rather than first time data directory selection.
* For the bitcoin.conf file, my thoughts were that if the user wanted to run side-by-side cash and non-cash they would have taken care of this as part of the setup for the alternate data directory.

Additionally, I had only intended this PR to deal with the Qt settings, so this will have no affect on using bitcoind rather than bitcoin-qt, whereas handling things with the bitcoin.conf file would affect both modes.

ptschip commented:
i thought by disconnecting we'd run into fewer problems, just wanted to
be sure that we don't cause or create any bans somehow and  it only take
a short while before those connections are re-established after the scan
is complete.

On 17/10/2017 12:12 PM, Justaphf wrote:
>
> @ptschip <https://github.com/ptschip> Thanks for taking a look at
> this. I'll rebase on dev and fix the boolean nit tonight or tomorrow.
>
> For the other issues
>
>   * As part of this PR I haven't touched the information displayed on
>     the select data dir dialog (just caused it to display again if
>     migrating from non-cash to cash). It's probably a good idea to
>     change that message, but seems like a separate PR unless we want
>     to add wording specific to migration rather than first time data
>     directory selection.
>   * For the bitcoin.conf file, my thoughts were that if the user
>     wanted to run side-by-side cash and non-cash they would have taken
>     care of this as part of the setup for the alternate data directory.
>
> Additionally, I had only intended this PR to deal with the Qt
> settings, so this will have no affect on using bitcoind rather than
> bitcoin-qt, whereas handling things with the bitcoin.conf file would
> affect both modes.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/756#issuecomment-337338714>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0LS3WNmXH48ZslJ6TuIv-LMU7D9Wks5stPwvgaJpZM4O7HFd>.
>

Justaphf commented:
Rebased on `dev` and fixed boolean variable naming nit.

ptschip commented:
@Justaphf the problem with the no-wallet travis issue is fixed now...just rebase on the new dev and you should be ok.

Justaphf committed:
Formatting

Justaphf commented:
Rebased and corrected merge conflicts.
Applied formatting fixes now that these files are part of format monitoring.

gandrewstone commented:
looks good, thanks!
